Automattic/jetpack
# WordPress.com Features: Calypso Locale Sync from wp-admin to Calypso

Related https://github.com/Automattic/jetpack/pull/37352 https://github.com/Automattic/jetpack/pull/19525 pfsHM7-Va-p2




Fixes https://github.com/Automattic/dotcom-forge/issues/6812

## Proposed changes:
Calypso Locale Sync from wp-admin to Calypso:

* Use the hook `profile_update` to get the wp-admin user profile locale update
* The locale is received in the format `en_US`, it's formatted as a language code `en-us`, and when that code doesn't exist, the language variation is removed `en`.
* When the user selects "Site Default", the locale is coming from the option `WPLANG`
* Request to `/wpcom/v2/me/locale` to update Calypso locale (user meta `locale` in wpcom database)
* When the user changes the "Site language" from `/wp-admin/options-general.php` and "Site Default" option is selected as the user locale, the hook `update_option_WPLANG` triggers the Calypso locale sync to the new Site language.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Test with the API from D148231-code
* Go to `/wp-admin/profile.php` on your Atomic test site, change and save the language.
  -  Try also with language variations such as "English (UK)" or "Español de Mexico"
![image](https://github.com/Automattic/jetpack/assets/1881481/3428546a-3fdd-4e97-83a9-2a28d04b620a)

* You should see the language change reflected on https://wordpress.com/sites
  - Verify the language is correctly selected in the dropdown at https://wordpress.com/me/account 

* Repeat the test, but this time choose the option "Site Default", which will use the Site language selected in `/wp-admin/options-general.php`

* Leave selected the option "Site Default" and change the "Site language" from `/wp-admin/options-general.php` to verify the Calypso locale is updated to the new Site language.

|wp-admin|Calypso|
|-|-|
|<img width="520" alt="Screenshot 2567-05-10 at 18 31 38" src="https://github.com/Automattic/jetpack/assets/1881481/565e2428-03e9-4814-83e0-f880e9eb4388">|<img width="362" alt="Screenshot 2567-05-10 at 18 31 28" src="https://github.com/Automattic/jetpack/assets/1881481/0311b343-2025-49ef-9a89-0c3aeaee9735">|
|<img width="540" alt="Screenshot 2567-05-10 at 18 32 16" src="https://github.com/Automattic/jetpack/assets/1881481/6987253f-ea26-424f-806c-8fef1365a78b">|<img width="366" alt="Screenshot 2567-05-10 at 18 32 09" src="https://github.com/Automattic/jetpack/assets/1881481/c828bc09-74bd-4c40-a25e-486c080c18e9">|
|<img width="522" alt="Screenshot 2567-05-10 at 19 14 20" src="https://github.com/Automattic/jetpack/assets/1881481/44ad578a-4149-4fc4-8a17-dadbfe998a23"><img width="525" alt="Screenshot 2567-05-10 at 19 09 15" src="https://github.com/Automattic/jetpack/assets/1881481/4afb8827-18a5-4260-a045-1190f73dd1a5">|<img width="362" alt="Screenshot 2567-05-10 at 18 31 28" src="https://github.com/Automattic/jetpack/assets/1881481/0311b343-2025-49ef-9a89-0c3aeaee9735">|
|<img width="510" alt="Screenshot 2567-05-14 at 18 54 58" src="https://github.com/Automattic/jetpack/assets/1881481/813fe8b6-6f09-457f-874d-1e26ec9ba909"><img width="515" alt="Screenshot 2567-05-14 at 18 59 05" src="https://github.com/Automattic/jetpack/assets/1881481/ccd59339-5098-4278-bb10-d78fce8ae248">|<img width="350" alt="Screenshot 2567-05-14 at 18 58 39" src="https://github.com/Automattic/jetpack/assets/1881481/fcd60aee-f5a7-48c7-b1d7-f9d7fb8c52f8">|
